File: scripts/recover_agent_transcripts.mjs
import fs from "node:fs";
import path from "node:path";
import os from "node:os";

const root = path.join(os.homedir(), ".cursor", "projects");
const outDir = path.join(process.cwd(), "cursor_chat_export", "recovered_from_agent_transcripts");
fs.mkdirSync(outDir, { recursive: true });

function cleanName(name, maxLen = 90) {
  const cleaned = name.replace(/[\\/:*?"<>|]+/g, "_").trim().replace(/\.+$/, "");
  return (cleaned || "untitled").slice(0, maxLen);
}

function extractText(content) {
  if (typeof content === "string") return content.trim();
  if (Array.isArray(content)) {
    const parts = [];
    for (const item of content) {
      if (typeof item === "string") parts.push(item);
      else if (item && typeof item === "object" && typeof item.text === "string") parts.push(item.text);
    }
    return parts.join("\n").trim();
  }
  if (content && typeof content === "object") {
    for (const k of ["text", "content", "message"]) {
      if (typeof content[k] === "string" && content[k].trim()) return content[k].trim();
    }
  }
  return "";
}

function bestTitle(firstUser, fallback) {
  if (!firstUser) return fallback;
  const firstLine = firstUser.split(/\r?\n/)[0] || "";
  const t = firstLine.replace("<user_query>", "").replace("</user_query>", "").trim();
  return t || fallback;
}

function renderMarkdown(records, transcriptId) {
  const messages = [];
  let firstUser = "";

  for (const rec of records) {
    const role = String(rec?.role || "").toLowerCase();
    const content = rec?.message?.content ?? [];
    let text = extractText(content);
    if (!text) continue;
    if (role === "assistant" && text.includes("[REDACTED]")) {
      text = text.replaceAll("[REDACTED]", "").trim();
      if (!text) continue;
    }
    if (role === "user" && !firstUser) firstUser = text;
    const label = role === "user" ? "User" : role === "assistant" ? "Assistant" : (role || "Message");
    messages.push({ label, text });
  }

  const title = bestTitle(firstUser, transcriptId.slice(0, 8));
  const lines = [`# ${title}`, "", `> transcript_id: \`${transcriptId}\``, "", "---", ""];
  for (const m of messages) lines.push(`### ${m.label}`, "", m.text, "", "---", "");
  return { title, markdown: `${lines.join("\n").trim()}\n` };
}

function walk(dir, out = []) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const ent of entries) {
    const full = path.join(dir, ent.name);
    if (ent.isDirectory()) walk(full, out);
    else if (ent.isFile() && ent.name.endsWith(".jsonl") && !full.includes(`${path.sep}subagents${path.sep}`)) out.push(full);
  }
  return out;
}

let recovered = 0;
let skipped = 0;
const files = fs.existsSync(root) ? walk(root) : [];
for (const file of files) {
  const text = fs.readFileSync(file, "utf8");
  const rows = [];
  for (const line of text.split(/\r?\n/)) {
    if (!line.trim()) continue;
    try {
      const obj = JSON.parse(line);
      if (obj && typeof obj === "object") rows.push(obj);
    } catch {}
  }
  if (rows.length === 0) {
    skipped += 1;
    continue;
  }
  const transcriptId = path.basename(file, ".jsonl");
  const { title, markdown } = renderMarkdown(rows, transcriptId);
  if (!markdown.includes("### ")) {
    skipped += 1;
    continue;
  }
  const outPath = path.join(outDir, `${cleanName(title)}__${transcriptId}.md`);
  fs.writeFileSync(outPath, markdown, "utf8");
  recovered += 1;
}

console.log(`Recovered: ${recovered}`);
console.log(`Skipped: ${skipped}`);
console.log(`Output: ${outDir}`);

File: scripts/recover_agent_transcripts.py
"""
Recover readable chat markdown files from local Cursor agent-transcripts JSONL logs.
"""
from __future__ import annotations

import json
import re
from pathlib import Path


ROOT = Path.home() / ".cursor" / "projects"
OUT_DIR = Path("cursor_chat_export") / "recovered_from_agent_transcripts"
OUT_DIR.mkdir(parents=True, exist_ok=True)


def clean_name(name: str, max_len: int = 90) -> str:
    name = re.sub(r'[\\/:*?"<>|]+', "_", name).strip().strip(".")
    return (name or "untitled")[:max_len]


def extract_text(content: object) -> str:
    if isinstance(content, str):
        return content.strip()
    if isinstance(content, list):
        chunks: list[str] = []
        for item in content:
            if isinstance(item, str):
                chunks.append(item)
            elif isinstance(item, dict):
                t = item.get("text")
                if isinstance(t, str) and t.strip():
                    chunks.append(t.strip())
        return "\n".join(chunks).strip()
    if isinstance(content, dict):
        for key in ("text", "content", "message"):
            value = content.get(key)
            if isinstance(value, str) and value.strip():
                return value.strip()
    return ""


def best_title(first_user_text: str, fallback: str) -> str:
    if not first_user_text:
        return fallback
    title = first_user_text.strip().splitlines()[0]
    title = title.replace("<user_query>", "").replace("</user_query>", "").strip()
    return title or fallback


def render_markdown(records: list[dict], transcript_id: str) -> tuple[str, str]:
    messages: list[tuple[str, str]] = []
    first_user = ""

    for rec in records:
        role = str(rec.get("role", "")).strip().lower()
        msg = rec.get("message", {})
        content = msg.get("content", []) if isinstance(msg, dict) else []
        text = extract_text(content)
        if not text:
            continue
        if role == "assistant" and "[REDACTED]" in text:
            text = text.replace("[REDACTED]", "").strip()
            if not text:
                continue
        if role == "user" and not first_user:
            first_user = text
        label = "User" if role == "user" else ("Assistant" if role == "assistant" else role.title() or "Message")
        messages.append((label, text))

    title = best_title(first_user, transcript_id[:8])
    lines = [f"# {title}", "", f"> transcript_id: `{transcript_id}`", "", "---", ""]
    for label, text in messages:
        lines.extend((f"### {label}", "", text, "", "---", ""))
    return title, "\n".join(lines).strip() + "\n"


def main() -> None:
    transcript_files = sorted(
        p for p in ROOT.glob("**/agent-transcripts/**/*.jsonl") if "\\subagents\\" not in str(p)
    )
    exported = 0
    skipped = 0

    for path in transcript_files:
        try:
            raw_lines = path.read_text(encoding="utf-8", errors="replace").splitlines()
        except Exception:
            skipped += 1
            continue
        rows: list[dict] = []
        for line in raw_lines:
            line = line.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
            except Exception:
                continue
            if isinstance(obj, dict):
                rows.append(obj)
        if not rows:
            skipped += 1
            continue

        transcript_id = path.stem
        title, md = render_markdown(rows, transcript_id)
        if "### " not in md:
            skipped += 1
            continue

        out_name = f"{clean_name(title)}__{transcript_id}.md"
        out_path = OUT_DIR / out_name
        out_path.write_text(md, encoding="utf-8")
        exported += 1

    print(f"Recovered: {exported}")
    print(f"Skipped: {skipped}")
    print(f"Output: {OUT_DIR.resolve()}")


if __name__ == "__main__":
    main()

